{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8cffa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap\n",
    "from matplotlib.patches import Ellipse\n",
    "import matplotlib.transforms as transforms\n",
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import seaborn as sns\n",
    "from datetime import date\n",
    "#set date for fig names\n",
    "today = date.today()\n",
    "date = today.strftime('%m%d%y')\n",
    "print(date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aebfac21",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define paths\n",
    "basedir='eclarki_genome_paper/'\n",
    "gene_exp_dir=basedir+'06_Gene_expression/'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19f8c374",
   "metadata": {},
   "outputs": [],
   "source": [
    "#paths to gene expression file and sample-condition file\n",
    "expFile=gene_exp_dir+'lengthScaledTPM.matrix' #normalized gene expression matrix (input file)\n",
    "sampleConditionFile=gene_exp_dir+'samples_conditions.txt' #sample-condition file groups samples for PCA\n",
    "! head {sampleConditionFile} #check input file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86c26b72",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PCA of all samples (Bryopsis and Elysia)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa85ce4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#link samples\n",
    "sampleConditionDict = {}\n",
    "\n",
    "for line in open(sampleConditionFile, 'r'):\n",
    "    line=line.rstrip().split('\\t')\n",
    "    if 'sample' in line[0]: #skip header\n",
    "        continue\n",
    "    sample,condition=line[0],line[1] #get samples, conditions from columns 1 and 2\n",
    "    sampleConditionDict[sample]=condition #store in dictionary with sample as key\n",
    "\n",
    "print('%i sample-condition pairs stored in sampleConditionDict\\n' % len(sampleConditionDict.keys()))\n",
    "\n",
    "df=pd.read_csv(expFile, header=0, sep='\\t').transpose() #bring in gene expression matrix and transpose\n",
    "display(df) #check dataframe\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd8cf8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples,conditions=[],[] #store list of all samples and conditions\n",
    "\n",
    "for sample in df.index: #go through each sample in df\n",
    "    if sample not in sampleConditionDict.keys(): #if sample not in samples/conditions file\n",
    "        df.drop([sample],inplace=True) #delete from the matrix\n",
    "        continue\n",
    "    condition=sampleConditionDict[sample] #otherwise, store it\n",
    "    conditions.append(condition)\n",
    "    samples.append(sample)\n",
    "\n",
    "print('%i unique conditions:' % len(set(conditions))) #check number of conditions is 16\n",
    "print('\\t'.join(set(conditions))+'\\n') #print conditions\n",
    "features=np.asarray(df.columns) #array of all genes in the dataframe\n",
    "features\n",
    "print('%i unique genes' % len(set(features)))\n",
    "\n",
    "#standardscale dataframe\n",
    "x=df.loc[:, features].values #store features (genes) as x\n",
    "x=StandardScaler().fit_transform(x) #normalize features\n",
    "x.shape #check shape of dataframe- first value=number of samples, second value=number of genes in matrix\n",
    "\n",
    "#check if the mean=0 and std=1\n",
    "print('mean: %s (%f)\\nstandard deviation: %s (%f)' % (np.mean(x),round(np.mean(x)),np.std(x),round(np.std(x))))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c87135cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat_cols=['gene'+str(i) for i in range(x.shape[1])] #rename columns to 'gene'+number\n",
    "\n",
    "normalised_df=pd.DataFrame(x,columns=feat_cols) #convert normalized data to a dataframe\n",
    "display(normalised_df) #check dataframe\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0785968c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.reset_index(inplace=True, drop=True) #change index to numbers\n",
    "df.head() #check dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f49ec26e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run PCA with n dimensions- PCA with all components\n",
    "\n",
    "n_components=15 #number of PCs\n",
    "\n",
    "columns=[]\n",
    "for i in range(n_components): #add principal componenets to columns\n",
    "    columns.append('principal component %s' % str(i+1))\n",
    "\n",
    "pca_comp=PCA(n_components=n_components) \n",
    "principalComponents_comp=pca_comp.fit_transform(x) #transform PCA variables\n",
    "print('Explained variation per principal component: {}'.format(pca_comp.explained_variance_ratio_)) #explained variance\n",
    "\n",
    "principal_Df=pd.DataFrame(data=principalComponents_comp, columns=columns) #store outputs as df for plotting\n",
    "df['label']=conditions #add label column\n",
    "df['sample']=samples #add sample column\n",
    "display(df.head())\n",
    "\n",
    "#build scree plot\n",
    "ind=np.arange(0, n_components) #range 1-n_components\n",
    "(fig, ax) = plt.subplots(figsize=(8, 6)) #define figure size\n",
    "sns.pointplot(x=ind, y=pca_comp.explained_variance_ratio_) #plot explained variance for each PC\n",
    "ax.set_title('Scree plot')\n",
    "ax.set_xticks(ind)\n",
    "ax.set_xticklabels(i+1 for i in ind) #correct PC number\n",
    "ax.set_xlabel('Component Number')\n",
    "ax.set_ylabel('Explained Variance')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1869f8ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plotting settings\n",
    "# print(set(conditions)) #use this to get list for targets (unordered)\n",
    "sns.set_theme(style='white') #set background to white\n",
    "\n",
    "targets=['Bp','E','V','L','AS'] #conditions in order (so they look best in fig legend)\n",
    "\n",
    "key_dict={'Bp':'Bryopsis','E':'egg','V':'veliger','L':'crawling larvae','AS':'juvenile slugs'}\n",
    "#Veliger\tBryopsis\tEgg\tJuvenile_slug\tLarval_slug\n",
    "\n",
    "name_dict={'#bdbdbd':'Bryopsis','#edf8fb':'   egg','#99d8c9':' veliger',\n",
    "           '#41ae76':'   larvae','#005824':'  adult'} #dict of hexcolor:condition name\n",
    "\n",
    "palette={'Bp':'#bdbdbd','E':'#edf8fb', 'V':'#99d8c9','L':'#41ae76','AS':'#005824'} #corresponding colors to each condition\n",
    "\n",
    "pallist=list(palette.values())\n",
    "sns.palplot(list(palette.values())) #display colors in palette\n",
    "ax=plt.gca()\n",
    "#add names of color/group\n",
    "for i, name in enumerate(pallist):\n",
    "#     print(i,name_dict[name])\n",
    "    ax.text(i-.4, -0.8, name_dict[name], fontsize=10) \n",
    "plt.show()\n",
    "\n",
    "markerdict={'B':'*','V':'v','L':'D','E':'o','A':'s'} #default assign markers to each condition\n",
    "\n",
    "outdir=''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dadc906",
   "metadata": {},
   "outputs": [],
   "source": [
    "def confidence_ellipse(x, y, ax, n_std=1.5, facecolor='none', **kwargs):\n",
    "    if x.size!=y.size:\n",
    "        print('x and y must be the same size')\n",
    "\n",
    "    cov=np.cov(x, y) #get covariance\n",
    "    pearson=cov[0, 1]/np.sqrt(cov[0, 0] * cov[1, 1])\n",
    "    #get eigenvalues of two-dimensionl dataset\n",
    "    ell_radius_x=np.sqrt(1 + pearson)\n",
    "    ell_radius_y=np.sqrt(1 - pearson)\n",
    "    ellipse=Ellipse((0, 0), width=ell_radius_x * 2, height=ell_radius_y * 2, facecolor=facecolor, **kwargs)\n",
    "\n",
    "    # calculate standard deviation of x from the squareroot of the variance and multiplying\n",
    "    # by given number of standard deviations\n",
    "    scale_x=np.sqrt(cov[0, 0]) * n_std\n",
    "    mean_x=np.mean(x)\n",
    "\n",
    "    # calculate standard deviation of y\n",
    "    scale_y=np.sqrt(cov[1, 1]) * n_std\n",
    "    mean_y=np.mean(y)\n",
    "    transf=transforms.Affine2D().rotate_deg(45).scale(scale_x, scale_y).translate(mean_x, mean_y)\n",
    "    ellipse.set_transform(transf + ax.transData)\n",
    "    return ax.add_patch(ellipse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "472cd9ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_pcs(targets,palette,df,principal_Df,keep_conditions,markerdict):\n",
    "\n",
    "    #dict defining PC variables for each plot: [1st PC (pca), 2nd PC (pcb), pca location in variance list, pcb location in variance list]\n",
    "    plotdict={'PC1 v PC2':['1','2',0,1],\n",
    "              'PC2 v PC3':['2','3',1,2],\n",
    "             'PC3 v PC4': ['3','4',2,3]}\n",
    "\n",
    "    # plotting PCs\n",
    "    for key in plotdict.keys():\n",
    "        print('plotting %s' % key)\n",
    "        fig=plt.figure(figsize=(10,10))\n",
    "        ax=plt.subplot(111)\n",
    "        plt.xticks(fontsize=12)\n",
    "        plt.yticks(fontsize=12)\n",
    "\n",
    "        custom_legend_data = []\n",
    "        for target in targets:\n",
    "            if target[0] in keep_conditions: #specify which conditions to keep for plot\n",
    "                color=palette[target]\n",
    "                markerid=target[0]\n",
    "                custom_legend_data.append(Line2D([0], [0], marker=markerdict[target[0]], color='w', label=key_dict[target],\n",
    "                                                 markerfacecolor=color, markersize=10, markeredgewidth=0.8, \n",
    "                                                 markeredgecolor='gray')) #build custom legend\n",
    "                indicesToKeep=df.index[df['label']==target].tolist() #find index in array of the condition\n",
    "\n",
    "                #plot scatterplot of just those indices (aka samples)\n",
    "                pca='principal component '+plotdict[key][0]\n",
    "                pcb='principal component '+plotdict[key][1]\n",
    "\n",
    "                plt.scatter(principal_Df.loc[indicesToKeep, pca], principal_Df.loc[indicesToKeep, pcb],\n",
    "                            s=100, label=target, marker=markerdict[target[0]], c=color, alpha=.8,\n",
    "                            edgecolor='gray', linewidth=0.8)\n",
    "                #plot ellipse around each condition\n",
    "                if len(keep_conditions)<3: #if displaying <3 conditions, make ellipses filled with color\n",
    "                    confidence_ellipse(principal_Df.loc[indicesToKeep, pca], principal_Df.loc[indicesToKeep,pcb],\n",
    "                                       ax, facecolor=color, edgecolor='gray', linewidth=2, alpha=.3, zorder=0)\n",
    "                else: #if all conditions, only show outline with color\n",
    "                    confidence_ellipse(principal_Df.loc[indicesToKeep, pca], principal_Df.loc[indicesToKeep, pcb],\n",
    "                                       ax, edgecolor=color, linewidth=3, alpha=.8, zorder=0)\n",
    "\n",
    "        box=ax.get_position()#shrink axis height by 10% on bottom\n",
    "        ax.set_position([box.x0, box.y0 + box.height * 0.1, box.width, box.height * 0.9]) \n",
    "\n",
    "        legend_cols=len(keep_conditions)\n",
    "        ax.legend(handles=custom_legend_data, loc='upper center', bbox_to_anchor=(0.5, -0.1),\n",
    "                 fancybox=True, ncol=legend_cols) #put legend below current axis \n",
    "\n",
    "        #get variance explained for each PC\n",
    "        pcva=plotdict[key][2]\n",
    "        pcvb=plotdict[key][3]\n",
    "        variance_pca,variance_pcb=pca_comp.explained_variance_ratio_[pcva]*100, pca_comp.explained_variance_ratio_[pcvb]*100\n",
    "\n",
    "        #plot PC name and variance explained on the axes\n",
    "        plt.xlabel('%s\\n(%f %%)' % (pca,variance_pca), fontsize=20)\n",
    "        plt.ylabel('%s\\n(%f %%)' % (pcb,variance_pcb), fontsize=20)\n",
    "\n",
    "        #make axes fit all datapoints and make them square\n",
    "        maxpca=max(principal_Df[pca]) #axes are 0+/-n where n is maximum PC of the principal components being plotted\n",
    "        maxpcb=max(principal_Df[pcb])\n",
    "        max_axis_size = max(maxpca, maxpcb) #take the larger of the two axes\n",
    "        lim=math.ceil(max_axis_size/100)*100 #round axis up (ie ceiling) length to nearest 100\n",
    "        print('max_axis_size',max_axis_size)\n",
    "        print('lim',lim)\n",
    "        plt.xlim(-lim,lim)\n",
    "        plt.ylim(-lim,lim)\n",
    "        plt.tight_layout()\n",
    "\n",
    "        keyname=key.replace(' ','') #remove spaces from key so that can be included in output figure name\n",
    "        conds=''\n",
    "        for x in keep_conditions: #make string of conditions to include in save name\n",
    "            conds+=x\n",
    "\n",
    "#         plt.savefig('%sElysia_Bryopsis_PC%svsPC%s.svg' % (outdir,pca[-1],pcb[-1]), format='svg', bbox_inches='tight')\n",
    "        print('plot saved as: %sElysia_Bryopsis_PC%svsPC%s.svg' % (outdir,pca[-1],pcb[-1]))\n",
    "        plt.show()\n",
    "\n",
    "        \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e11cc0b1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "keep_conditions=['B','E','V','L','A'] #conditions to include\n",
    "\n",
    "plot_pcs(targets,palette,df,principal_Df,keep_conditions,markerdict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e13ec83d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73e5bdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PCA of with only Elysia samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bde29d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#link samples\n",
    "sampleConditionDict = {}\n",
    "\n",
    "for line in open(sampleConditionFile, 'r'):\n",
    "    line=line.rstrip().split('\\t')\n",
    "    if 'sample' in line[0]: #skip header\n",
    "        continue\n",
    "    sample,condition=line[0],line[1] #get samples, conditions from columns 1 and 2\n",
    "    if 'Bp' not in sample: #skip Bryopsis\n",
    "        sampleConditionDict[sample]=condition #store in dictionary with sample as key\n",
    "\n",
    "print('%i sample-condition pairs stored in sampleConditionDict\\n' % len(sampleConditionDict.keys()))\n",
    "sampleConditionDict\n",
    "df=pd.read_csv(expFile, header=0, sep='\\t').transpose() #bring in gene expression matrix and transpose\n",
    "df=df[~df.index.str.startswith('Bp')] #remove bryopsis \n",
    "\n",
    "df=df.loc[:,(df!=0).any(axis=0)] #drop columns with all 0s-- fixes issue with standard dev in next block\n",
    "\n",
    "display(df) #check dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae0dfdfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples,conditions=[],[] #store list of all samples and conditions\n",
    "\n",
    "for sample in df.index: #go through each sample in df\n",
    "    if sample not in sampleConditionDict.keys(): #if sample not in samples/conditions file\n",
    "        df.drop([sample],inplace=True) #delete from the matrix\n",
    "        continue\n",
    "    condition=sampleConditionDict[sample] #otherwise, store it\n",
    "    conditions.append(condition)\n",
    "    samples.append(sample)\n",
    "\n",
    "print('%i unique conditions:' % len(set(conditions))) #check number of conditions is 16\n",
    "print('\\t'.join(set(conditions))+'\\n') #print conditions\n",
    "features=np.asarray(df.columns) #array of all genes in the dataframe\n",
    "features\n",
    "print('%i unique genes' % len(set(features)))\n",
    "\n",
    "#standardscale dataframe\n",
    "x=df.loc[:, features].values #store features (genes) as x\n",
    "x=StandardScaler().fit_transform(x) #normalize features\n",
    "x.shape #check shape of dataframe- first value=number of samples, second value=number of genes in matrix\n",
    "\n",
    "#check if the mean=0 and std=1\n",
    "print('mean: %s (%f)\\tstandard deviation: %s (%f)' % (np.mean(x),round(np.mean(x)),np.std(x),round(np.std(x))))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78659112",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat_cols=['gene'+str(i) for i in range(x.shape[1])] #rename columns to 'gene'+number\n",
    "\n",
    "normalised_df=pd.DataFrame(x,columns=feat_cols) #convert normalized data to a dataframe\n",
    "display(normalised_df) #check dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed59b2b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.reset_index(inplace=True, drop=True) #change index to numbers\n",
    "df.head() #check dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ddc1c43",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run PCA with n dimensions- PCA with all components\n",
    "\n",
    "n_components=12 #number of PCs\n",
    "\n",
    "columns=[]\n",
    "for i in range(n_components): #add principal componenets to columns\n",
    "    columns.append('principal component %s' % str(i+1))\n",
    "\n",
    "pca_comp=PCA(n_components=n_components) \n",
    "principalComponents_comp=pca_comp.fit_transform(x) #transform PCA variables\n",
    "print('Explained variation per principal component: {}'.format(pca_comp.explained_variance_ratio_)) #explained variance\n",
    "\n",
    "principal_Df=pd.DataFrame(data=principalComponents_comp, columns=columns) #store outputs as df for plotting\n",
    "df['label']=conditions #add label column\n",
    "df['sample']=samples #add sample column\n",
    "display(df.head())\n",
    "\n",
    "#build scree plot\n",
    "ind=np.arange(0, n_components) #range 1-n_components\n",
    "(fig, ax) = plt.subplots(figsize=(8, 6)) #define figure size\n",
    "sns.pointplot(x=ind, y=pca_comp.explained_variance_ratio_) #plot explained variance for each PC\n",
    "ax.set_title('Scree plot')\n",
    "ax.set_xticks(ind)\n",
    "ax.set_xticklabels(i+1 for i in ind) #correct PC number\n",
    "ax.set_xlabel('Component Number')\n",
    "ax.set_ylabel('Explained Variance')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9a46a37",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2c8efe1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Plotting settings\n",
    "# print(set(conditions)) #use this to get list for targets (unordered)\n",
    "sns.set_theme(style='white') #set background to white\n",
    "\n",
    "targets=['E','V','L','AS'] #conditions in order (so they look best in fig legend)\n",
    "\n",
    "key_dict={'Bp':'Bryopsis','E':'egg','V':'veliger','L':'larvae','AS':'adult'}\n",
    "\n",
    "name_dict={'#edf8fb':'   egg','#99d8c9':' veliger',\n",
    "           '#41ae76':'  larvae','#005824':'  adult'} #dict of hexcolor:condition name\n",
    "\n",
    "palette={'E':'#edf8fb', 'V':'#99d8c9','L':'#41ae76','AS':'#005824'} #corresponding colors to each condition\n",
    "\n",
    "pallist=list(palette.values())\n",
    "sns.palplot(list(palette.values())) #display colors in palette\n",
    "ax=plt.gca()\n",
    "#add names of color/group\n",
    "for i, name in enumerate(pallist):\n",
    "#     print(i,name_dict[name])\n",
    "    ax.text(i-.4, -0.8, name_dict[name], fontsize=10) \n",
    "plt.show()\n",
    "\n",
    "markerdict={'V':'v','L':'D','E':'o','A':'s'} #default assign markers to each condition\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd34ec3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def confidence_ellipse(x, y, ax, n_std=1.5, facecolor='none', **kwargs):\n",
    "    if x.size!=y.size:\n",
    "        print('x and y must be the same size')\n",
    "\n",
    "    cov=np.cov(x, y) #get covariance\n",
    "    pearson=cov[0, 1]/np.sqrt(cov[0, 0] * cov[1, 1])\n",
    "    #get eigenvalues of two-dimensionl dataset\n",
    "    ell_radius_x=np.sqrt(1 + pearson)\n",
    "    ell_radius_y=np.sqrt(1 - pearson)\n",
    "    ellipse=Ellipse((0, 0), width=ell_radius_x * 2, height=ell_radius_y * 2, facecolor=facecolor, **kwargs)\n",
    "\n",
    "    # calculate standard deviation of x from the squareroot of the variance and multiplying\n",
    "    # by given number of standard deviations\n",
    "    scale_x=np.sqrt(cov[0, 0]) * n_std\n",
    "    mean_x=np.mean(x)\n",
    "\n",
    "    # calculate standard deviation of y\n",
    "    scale_y=np.sqrt(cov[1, 1]) * n_std\n",
    "    mean_y=np.mean(y)\n",
    "    transf=transforms.Affine2D().rotate_deg(45).scale(scale_x, scale_y).translate(mean_x, mean_y)\n",
    "    ellipse.set_transform(transf + ax.transData)\n",
    "    return ax.add_patch(ellipse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "614bec24",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_pcs(targets,palette,df,principal_Df,keep_conditions,markerdict):\n",
    "\n",
    "    #dict defining PC variables for each plot: [1st PC (pca), 2nd PC (pcb), pca location in variance list, pcb location in variance list]\n",
    "    plotdict={'PC1 v PC2':['1','2',0,1],\n",
    "              'PC2 v PC3':['2','3',1,2],\n",
    "             'PC3 v PC4': ['3','4',2,3]}\n",
    "\n",
    "    # plotting PCs\n",
    "    for key in plotdict.keys():\n",
    "        print('plotting %s' % key)\n",
    "        fig=plt.figure(figsize=(10,10))\n",
    "        ax=plt.subplot(111)\n",
    "        plt.xticks(fontsize=12)\n",
    "        plt.yticks(fontsize=12)\n",
    "\n",
    "        custom_legend_data = []\n",
    "        for target in targets:\n",
    "            if target[0] in keep_conditions: #specify which conditions to keep for plot\n",
    "                color=palette[target]\n",
    "                markerid=target[0]\n",
    "                custom_legend_data.append(Line2D([0], [0], marker=markerdict[target[0]], color='w', label=key_dict[target],\n",
    "                                                 markerfacecolor=color, markersize=10, markeredgewidth=0.8, \n",
    "                                                 markeredgecolor='gray')) #build custom legend\n",
    "                indicesToKeep=df.index[df['label']==target].tolist() #find index in array of the condition\n",
    "\n",
    "                #plot scatterplot of just those indices (aka samples)\n",
    "                pca='principal component '+plotdict[key][0]\n",
    "                pcb='principal component '+plotdict[key][1]\n",
    "\n",
    "                plt.scatter(principal_Df.loc[indicesToKeep, pca], principal_Df.loc[indicesToKeep, pcb],\n",
    "                            s=100, label=target, marker=markerdict[target[0]], c=color, alpha=.8,\n",
    "                            edgecolor='gray', linewidth=0.8)\n",
    "                #plot ellipse around each condition\n",
    "                if len(keep_conditions)<3: #if displaying <3 conditions, make ellipses filled with color\n",
    "                    confidence_ellipse(principal_Df.loc[indicesToKeep, pca], principal_Df.loc[indicesToKeep,pcb],\n",
    "                                       ax, facecolor=color, edgecolor='gray', linewidth=2, alpha=.3, zorder=0)\n",
    "                else: #if all conditions, only show outline with color\n",
    "                    confidence_ellipse(principal_Df.loc[indicesToKeep, pca], principal_Df.loc[indicesToKeep, pcb],\n",
    "                                       ax, edgecolor=color, linewidth=3, alpha=.8, zorder=0)\n",
    "\n",
    "        box=ax.get_position()#shrink axis height by 10% on bottom\n",
    "        ax.set_position([box.x0, box.y0 + box.height * 0.1, box.width, box.height * 0.9]) \n",
    "\n",
    "        legend_cols=len(keep_conditions)\n",
    "        ax.legend(handles=custom_legend_data, loc='upper center', bbox_to_anchor=(0.5, -0.1),\n",
    "                 fancybox=True, ncol=legend_cols) #put legend below current axis \n",
    "\n",
    "        #get variance explained for each PC\n",
    "        pcva=plotdict[key][2]\n",
    "        pcvb=plotdict[key][3]\n",
    "        variance_pca,variance_pcb=pca_comp.explained_variance_ratio_[pcva]*100, pca_comp.explained_variance_ratio_[pcvb]*100\n",
    "\n",
    "        #plot PC name and variance explained on the axes\n",
    "        plt.xlabel('%s\\n(%f %%)' % (pca,variance_pca), fontsize=20)\n",
    "        plt.ylabel('%s\\n(%f %%)' % (pcb,variance_pcb), fontsize=20)\n",
    "\n",
    "        #make axes fit all datapoints and make them square\n",
    "        maxpca=max(principal_Df[pca]) #axes are 0+/-n where n is maximum PC of the principal components being plotted\n",
    "        maxpcb=max(principal_Df[pcb])\n",
    "        max_axis_size = max(maxpca, maxpcb) #take the larger of the two axes\n",
    "        lim=math.ceil(max_axis_size/100)*100+20 #round axis up (ie ceiling) length to nearest 100\n",
    "        print('max_axis_size',max_axis_size)\n",
    "        print('lim',lim)\n",
    "        plt.xlim(-lim,lim)\n",
    "        plt.ylim(-lim,lim)\n",
    "        plt.tight_layout()\n",
    "\n",
    "        keyname=key.replace(' ','') #remove spaces from key so that can be included in output figure name\n",
    "        conds=''\n",
    "        for x in keep_conditions: #make string of conditions to include in save name\n",
    "            conds+=x\n",
    "\n",
    "        plt.savefig('%sElysia_PC%svsPC%s.svg' % (outdir,pca[-1],pcb[-1]), format='svg', bbox_inches='tight')\n",
    "        print('%sElysia_PC%svsPC%s.svg' % (outdir,pca[-1],pcb[-1]))\n",
    "        \n",
    "        plt.show()\n",
    "\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48b1a157",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# keep_conditions=['E','V'] #conditions to include\n",
    "\n",
    "# plot_pcs(targets,palette,df,principal_Df,keep_conditions,markerdict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f43d13",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "keep_conditions=['E','V','L','A'] #conditions to include\n",
    "\n",
    "plot_pcs(targets,palette,df,principal_Df,keep_conditions,markerdict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fce3ff54",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
