{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3dbde609",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ['QT_QPA_PLATFORM']='offscreen'\n",
    "import re\n",
    "import Bio\n",
    "from Bio import Seq, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c9764475",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ete3\n",
    "from ete3 import Tree, faces, TreeStyle, PhyloTree, NodeStyle, TextFace, AttrFace, SeqMotifFace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b36a52ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "treefile = '../../../../figshare/pks_phylogeny/KS_domain_cdhit80_prune.mpr.tree'\n",
    "outfile = '../../../../figshare/pks_phylogeny/KS_domain_cdhit80_prune.mpr.pdf'\n",
    "alnfile = '../../../../figshare/pks_phylogeny/KS_domain_cdhit80_prune.aln'\n",
    "#scriptsdir = '.'\n",
    "colorscheme = 'metazoa'\n",
    "branch_min = 0.95\n",
    "\n",
    "nodefile = 'nodes.dmp'\n",
    "mergedfile = 'merged.dmp'\n",
    "lineages_colors_file = 'lineage_colors.dmp'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "606e7fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqDict = {}\n",
    "alnLen = 0\n",
    "for record in SeqIO.parse(alnfile, \"fasta\"):\n",
    "    name = record.id\n",
    "    sequence = str(record.seq)\n",
    "    alnLen = len(sequence)\n",
    "    seqDict[name] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f09dc2e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Initialize the taxonomy dictionary\n",
    "taxnodes = {}\n",
    "fi = open(nodefile)\n",
    "for line in fi:\n",
    "    line = line.rstrip('\\n').split('\\t')\n",
    "    node, parent = line[0], line[2]\n",
    "    taxnodes[node] = parent\n",
    "fi.close()\n",
    "\n",
    "fi = open(mergedfile)\n",
    "for line in fi:\n",
    "    line = line.rstrip('\\n').split('\\t')\n",
    "    node, newnode = line[0], line[2]\n",
    "    taxnodes[node] = taxnodes[newnode]\n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a52dd1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in the lineages colors file and store in dictionaries \n",
    "orderDict = {}\n",
    "colorDict = {}\n",
    "lineages = {}\n",
    "fi = open(lineages_colors_file)\n",
    "\n",
    "for line in fi:\n",
    "    line = line.rstrip('\\n').split('\\t')\n",
    "    \n",
    "    if line[3] == colorscheme:\n",
    "        order, taxid_list, name, color = line[0], line[5], line[4], line[2]\n",
    "        #print(order, name, color, taxid_list)\n",
    "        orderDict[int(order)] = name\n",
    "        colorDict[name] = color\n",
    "        \n",
    "        for taxid in taxid_list.split(','):\n",
    "            #print(taxid, name)\n",
    "            lineages[taxid] = name\n",
    "            \n",
    "fi.close()\n",
    "#print(colorDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3e2ed8f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create function for returning the lineage string for a given taxonomy id\n",
    "def taxdump(taxid):\n",
    "    root = ''\n",
    "    taxlist = []\n",
    "    \n",
    "    while root == '':\n",
    "        if taxid == '':\n",
    "            root = '1'\n",
    "        elif taxid == '1':\n",
    "            root = '1'\n",
    "        else: \n",
    "            taxlist.append(taxid)\n",
    "            taxid = taxnodes[taxid]\n",
    "            \n",
    "    return taxlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8eac69c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add species and lineage information to all leaves\n",
    "# color target gene red \n",
    "# color all ME034 gene green\n",
    "\n",
    "#t = PhyloTree(treefile, alignment=alnfile, alg_format=\"fasta\")\n",
    "t = Tree(treefile, format=1)\n",
    "t.ladderize(direction=1)\n",
    "leafSet = set()\n",
    "\n",
    "#print(\"Custom mode:\")\n",
    "for n in t.get_leaves():\n",
    "    leafSet.add(n.name)\n",
    "    #print(n.name)\n",
    "    speciesname = ''\n",
    "    genename = ''\n",
    "    taxid = ''\n",
    "    \n",
    "    tmp = n.name.split(\"-\")\n",
    "    speciesname = tmp[2]\n",
    "    speciesname = speciesname.replace(\"_\", \" \")\n",
    "    genename = tmp[0]\n",
    "    taxid = tmp[1]\n",
    "    #print(speciesname, genename, taxid)\n",
    "\n",
    "    lincolor = '#000000'\n",
    "    linname = ''\n",
    "    lineage = taxdump(taxid)\n",
    "    #print(lineage)\n",
    "    \n",
    "    for i in lineage:\n",
    "        if str(i) in lineages:\n",
    "            linname = lineages[str(i)]\n",
    "            lincolor = colorDict[linname]\n",
    "            #print(linname,lincolor)\n",
    "            break\n",
    "    \n",
    "    n.add_features(lineage=linname)\n",
    "    n.add_features(gene=genename)\n",
    "    n.add_features(species=speciesname)\n",
    "    n.add_features(taxid=taxid)\n",
    "\n",
    "    #print(\"Species name:\", n.species, \"Species lineage:\", n.lineage, \"Color:\", lincolor)\n",
    "\n",
    "    # create a new label with a color attribute\n",
    "    linF = AttrFace(\"lineage\", fgcolor=lincolor, fsize=1)\n",
    "    linF.background.color = lincolor\n",
    "    linF.margin_top = linF.margin_bottom = linF.margin_left = 10\n",
    "    \n",
    "    speciesF = AttrFace(\"species\", fsize=10, fgcolor=lincolor, fstyle=\"italic\")\n",
    "    speciesF.margin_right = speciesF.margin_left = 10\n",
    "    taxidF = AttrFace(\"taxid\", fsize=10, fgcolor=lincolor, fstyle=\"normal\")\n",
    "    taxidF.margin_right = taxidF.margin_left = 10\n",
    "\n",
    "    if speciesname == 'Elysia crispata':\n",
    "        geneF = AttrFace(\"gene\", fsize=12, fgcolor=\"red\", fstyle=\"bold\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "    \n",
    "    else:\n",
    "        geneF = AttrFace(\"gene\", fsize=10, fgcolor=\"black\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "\n",
    "    # labels aligned to the same level\n",
    "    n.add_face(speciesF, 0, position='aligned')\n",
    "    n.add_face(geneF, 0, position='branch-right')\n",
    "    n.add_face(taxidF, 1, position='aligned')\n",
    "    n.add_face(linF, 2, position='aligned')\n",
    "    \n",
    "    my_motifs = [[0, alnLen, \"compactseq\", 2, 10, None, None, None]]\n",
    "    seqF = SeqMotifFace(seq=seqDict[n.name], motifs=my_motifs, gap_format=\"blank\")\n",
    "    seqF.margin_right = seqF.margin_left = 5\n",
    "    n.add_face(seqF, 3, \"aligned\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "53e12b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add lineage information to all internal nodes\n",
    "style = NodeStyle()\n",
    "\n",
    "style[\"size\"] = 0\n",
    "style[\"hz_line_width\"] = 2\n",
    "style[\"vt_line_width\"] = 2\n",
    "t.set_style(style)\n",
    "\n",
    "for n in t.iter_descendants(\"postorder\"):\n",
    "    #print(n.name)\n",
    "                \n",
    "    style[\"size\"] = 0\n",
    "    style[\"hz_line_width\"] = 2\n",
    "    style[\"vt_line_width\"] = 2\n",
    "    n.set_style(style)\n",
    "    \n",
    "    lineage_set = set()\n",
    "    # get descendants, if all descendants are members of same lineage, color lineage color\n",
    "    #print(\"NODE CHILDREN:\")\n",
    "    for k in n.iter_descendants(\"postorder\"):\n",
    "        for l in k.get_leaves():\n",
    "            lineage_set.add(l.lineage)\n",
    "            #print(\"Gene:\", l.gene, \"Species:\", l.species, \"Lineage:\", l.lineage, \"Color:\", lin2color[l.lineage])\n",
    "    \n",
    "    #print(len(lineage_set), lineage_set)\n",
    "    if len(lineage_set) == 1:\n",
    "        node_lin = ''.join(lineage_set)\n",
    "        #print(len(lineage_set), lineage_set, node_lin)\n",
    "    \n",
    "        newstyle = NodeStyle()\n",
    "        newstyle[\"size\"] = 0\n",
    "        newstyle[\"hz_line_width\"] = 2\n",
    "        newstyle[\"vt_line_width\"] = 2\n",
    "        newstyle[\"vt_line_color\"] = colorDict[node_lin]\n",
    "        newstyle[\"hz_line_color\"] = colorDict[node_lin]\n",
    "        n.img_style = newstyle\n",
    "        \n",
    "    #fix branchlengths?\n",
    "    \n",
    "    if n.name not in leafSet and n.name[0] != 'n':\n",
    "        #print(n.name)\n",
    "        \n",
    "        if float(n.name) >= branch_min:\n",
    "            #branch_support = n.name.split('.')[0]\n",
    "            branch_support = n.name\n",
    "            #print(branch_support)\n",
    "            n.add_features(bootstrap=branch_support)\n",
    "            \n",
    "            if len(lineage_set) == 1:\n",
    "                node_lin = ''.join(lineage_set)\n",
    "                supF = AttrFace(\"bootstrap\", fgcolor=colorDict[node_lin], fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-bottom')\n",
    "                \n",
    "            else:\n",
    "                supF = AttrFace(\"bootstrap\", fgcolor=\"#000000\", fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-bottom')\n",
    "        \n",
    "\n",
    "for n in t.get_leaves():\n",
    "    \n",
    "    leafstyle = NodeStyle()\n",
    "    leafstyle[\"size\"] = 0\n",
    "    leafstyle[\"hz_line_width\"] = 2\n",
    "    leafstyle[\"vt_line_width\"] = 2\n",
    "    leafstyle[\"vt_line_color\"] = colorDict[n.lineage]\n",
    "    leafstyle[\"hz_line_color\"] = colorDict[n.lineage]\n",
    "    n.img_style = leafstyle\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5e87ebff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add legend\n",
    "\n",
    "ts = TreeStyle()\n",
    "ts.show_leaf_name = False\n",
    "#ts.show_branch_support = True\n",
    "ts.draw_guiding_lines = True\n",
    "\n",
    "ts.title.add_face(TextFace(\"Taxonomy:\", fsize=10), column=0)\n",
    "for i in range(1, len(colorDict)):\n",
    "    #print(orderDict[i], colorDict[orderDict[i]])\n",
    "    ts.title.add_face(TextFace(orderDict[i], fsize=10, fgcolor=colorDict[orderDict[i]]), column=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3d16346",
   "metadata": {},
   "outputs": [],
   "source": [
    "# render image on notebook or save to file\n",
    "t.render(outfile, tree_style=ts)\n",
    "#t.render(\"%%inline\", tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b39de418",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d19177ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
